src/libexpr/primops.cc

```diff
@@ -1043,6 +1041,48 @@ static RegisterPrimOp primop_trace({
     .fun = prim_trace,
 });
 
+static void prim_warn(EvalState & state, const PosIdx pos, Value * * args, Value & v)
+{
+    // We only accept a string argument for now. The use case for pretty printing a value is covered by `trace`.
+    // By rejecting non-strings we allow future versions to add more features without breaking existing code.
+    auto msgStr = state.forceString(*args[0], pos, "while evaluating the first argument; the message passed to builtins.warn");
+
+    {
+        BaseError msg(std::string{msgStr});
+        msg.atPos(state.positions[pos]);
+        auto info = msg.info();
+        info.level = lvlWarn;
+        logWarning(info);
```

Comment:
If I understand correctly, users will not be able to easily distinguish output produced by `builtins.warn` and output produced by other [calls of `logWarning`](https://github.com/search?q=repo%3ANixOS%2Fnix+logWarning&type=code). However, other callsites will not go ahead and trigger the "abort on warn" behaviour. It's a safe bet that you'll have some users set `NIX_ABORT_ON_WARN`, then hit one of the other codepaths, and be confuse why Nix did *not* abort.

I don't think that it's a solution to have Nix abort on all these kinds of warnings. There are rather benign warnings too, like:

```
warning: Git tree '...' is dirty
warning: unknown flake output 'homeModules'
```

The only proposal I have is to use some other logging prefix, like `eval-warning:` instead of `warning:` to make the distinction more clear.